library/std/src/sys/unix/thread_local_dtor.rs

```diff
@@ -16,23 +16,36 @@ pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
     use crate::mem;
     use crate::sys_common::thread_local_dtor::register_dtor_fallback;
 
+    /// This is necessary because the __cxa_thread_atexit_impl implementation
+    /// std links to by default may be a C or C++ implementation that was not
+    /// compiled using the Clang integer normalization option.
+    #[cfg(not(sanitizer_cfi_normalize_integers))]
+    #[cfi_encoding = "i"]
+    #[repr(transparent)]
+    pub struct c_int(pub libc::c_int);
+
     extern "C" {
         #[linkage = "extern_weak"]
         static __dso_handle: *mut u8;
         #[linkage = "extern_weak"]
-        static __cxa_thread_atexit_impl: *const libc::c_void;
+        fn __cxa_thread_atexit_impl(
+            dtor: unsafe extern "C" fn(*mut libc::c_void),
+            arg: *mut libc::c_void,
+            dso_handle: *mut libc::c_void,
+        ) -> c_int;
     }
-    if !__cxa_thread_atexit_impl.is_null() {
-        type F = unsafe extern "C" fn(
-            dtor: unsafe extern "C" fn(*mut u8),
-            arg: *mut u8,
-            dso_handle: *mut u8,
-        ) -> libc::c_int;
-        mem::transmute::<*const libc::c_void, F>(__cxa_thread_atexit_impl)(
-            dtor,
-            t,
-            &__dso_handle as *const _ as *mut _,
-        );
+
+    if !(__cxa_thread_atexit_impl as *mut libc::c_void).is_null() {
+        unsafe {
+            __cxa_thread_atexit_impl(
+                mem::transmute::<
+                    unsafe extern "C" fn(*mut u8),
+                    unsafe extern "C" fn(*mut libc::c_void),
+                >(dtor),
+                mem::transmute::<*mut u8, *mut libc::c_void>(t),
+                &__dso_handle as *const _ as *mut _,
```

Comment:
You want a pointer to the static, which is itself a pointer, right? I think this is clearer if you use
```suggestion
                ptr::addr_of_mut!(__dso_handle),
```

In this case I believe it is approximately the same, but it "goes directly to it" in a way that is easier to reason about in light of the many nuances of interacting with statics and borrows of them:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=34bca38f5fa3c2f6d158f46a3bfc7a37